File: Packages/Scoring/Sources/Scoring/MatchEngine.swift
import Foundation

/// The result of running one match (side A vs side B over an ordered set of
/// holes). Shared by Match Play, every Nassau segment, and every press.
public struct MatchComputation: Sendable {
    public struct HoleResult: Hashable, Sendable {
        public let hole: Int
        /// nil = halved.
        public let winner: MatchSide?
        /// Cumulative up-count (A positive) after this hole.
        public let upAAfter: Int
        /// True when the hole was decided by concession (withdrawal) rather
        /// than by scores.
        public let byConcession: Bool
    }

    public let status: MatchStatus
    /// Decided holes in play order (stops at the point the match closed).
    public let holeResults: [HoleResult]
    /// Holes with no result yet (missing scores), still able to count.
    public let pendingHoles: [Int]
}

/// Pure hole-by-hole match evaluator.
///
/// Rules encoded here:
/// - **Best ball**: a side's score on a hole is the lowest net score among its
///   active members. All active members of both sides must have posted before
///   the hole counts (a partially-entered best-ball hole could otherwise flip
///   after being "decided").
/// - **Pending holes stay live**: a skipped hole (no scores yet) counts toward
///   `remaining`, because it can still be filled in. Mathematical closure
///   (`|up| > remaining`) is therefore safe even with gaps: the trailing side
///   cannot catch up even by winning every pending hole.
/// - **Withdrawals concede**: once every member of a side has withdrawn, each
///   later hole is conceded to the other side (if it still has anyone
///   standing). If both sides are gone, remaining holes are halved.
/// - **Auto-close**: once `|up| > remaining` the match is over; later holes
///   are ignored for this match.
public enum MatchEngine {

    /// Who won each hole, with no notion of match closure. Nassau computes
    /// this once per segment and derives the original bet and every press
    /// from the same outcome table.
    public struct RawOutcome: Hashable, Sendable {
        public let winner: MatchSide?   // nil = halved
        public let byConcession: Bool
    }

    public struct RawOutcomes: Sendable {
        /// Decided holes only.
        public let byHole: [Int: RawOutcome]
        /// Holes with missing scores, ascending.
        public let pendingHoles: [Int]
    }

    public static func rawOutcomes(
        sideA: [PlayerID],
        sideB: [PlayerID],
        holes: [Int],
        snapshot: RoundSnapshot,
        mode: HandicapMode,
        allowance: HandicapAllowance
    ) -> RawOutcomes {
        let participants = (sideA + sideB).compactMap { snapshot.player($0) }
        let courseHoles = snapshot.holeNumbers.compactMap { snapshot.course.hole($0) }
        // Strokes fall on the holes where the full round's stroke index puts
        // them — a front-nine segment does not re-spread the handicap.
        let strokes = StrokeAllocator.table(
            for: participants,
            holes: courseHoles,
            mode: mode,
            allowance: allowance
        )

        func net(_ player: PlayerID, _ hole: Int) -> Int? {
            guard let gross = snapshot.gross(player, hole: hole) else { return nil }
            return gross - (strokes[player]?[hole] ?? 0)
        }

        var byHole: [Int: RawOutcome] = [:]
        var pending: [Int] = []

        for hole in holes.sorted() {
            let activeA = snapshot.activePlayers(of: sideA, atHole: hole)
            let activeB = snapshot.activePlayers(of: sideB, atHole: hole)

            switch (activeA.isEmpty, activeB.isEmpty) {
            case (true, true):
                byHole[hole] = RawOutcome(winner: nil, byConcession: true)
            case (true, false):
                byHole[hole] = RawOutcome(winner: .b, byConcession: true)
            case (false, true):
                byHole[hole] = RawOutcome(winner: .a, byConcession: true)
            case (false, false):
                let netsA = activeA.map { net($0, hole) }
                let netsB = activeB.map { net($0, hole) }
                if netsA.contains(where: { $0 == nil }) || netsB.contains(where: { $0 == nil }) {
                    pending.append(hole)
                } else {
                    let bestA = netsA.compactMap { $0 }.min()!
                    let bestB = netsB.compactMap { $0 }.min()!
                    let winner: MatchSide? = bestA < bestB ? .a : (bestB < bestA ? .b : nil)
                    byHole[hole] = RawOutcome(winner: winner, byConcession: false)
                }
            }
        }

        return RawOutcomes(byHole: byHole, pendingHoles: pending)
    }

    /// Applies cumulative match logic (closure, dormie, display) to a hole
    /// range using a precomputed outcome table.
    public static func status(over holes: [Int], outcomes: RawOutcomes) -> MatchComputation {
        var results: [MatchComputation.HoleResult] = []
        var pending: [Int] = []
        var upA = 0
        var closed = false

        for hole in holes.sorted() {
            if closed { break }
            guard let outcome = outcomes.byHole[hole] else {
                pending.append(hole)
                continue
            }
            switch outcome.winner {
            case .a: upA += 1
            case .b: upA -= 1
            case nil: break
            }
            results.append(.init(hole: hole, winner: outcome.winner, upAAfter: upA, byConcession: outcome.byConcession))

            // Undecided = total - decided. Pending holes are still undecided
            // (they may be filled in later), so they count toward what the
            // trailing side could still win.
            let undecided = holes.count - results.count
            if abs(upA) > undecided {
                closed = true
            }
        }

        let undecided = holes.count - results.count

        var status = MatchStatus(
            holesDecided: results.count,
            upA: upA,
            remaining: undecided,
            closed: closed,
            winner: nil,
            dormieSide: nil,
            thruHole: results.map(\.hole).max()
        )

        if closed {
            status.winner = upA > 0 ? .a : .b
        } else if undecided == 0 {
            status.winner = upA > 0 ? .a : (upA < 0 ? .b : nil)
        } else if abs(upA) == undecided, upA != 0 {
            status.dormieSide = upA > 0 ? .a : .b
        }

        status.display = display(for: status)

        return MatchComputation(status: status, holeResults: results, pendingHoles: pending)
    }

    /// One-shot convenience: raw outcomes + status over the same range.
    public static func compute(
        sideA: [PlayerID],
        sideB: [PlayerID],
        holes: [Int],
        snapshot: RoundSnapshot,
        mode: HandicapMode,
        allowance: HandicapAllowance
    ) -> MatchComputation {
        let outcomes = rawOutcomes(
            sideA: sideA, sideB: sideB, holes: holes,
            snapshot: snapshot, mode: mode, allowance: allowance
        )
        return status(over: holes, outcomes: outcomes)
    }

    /// Conventional match-play status strings.
    static func display(for status: MatchStatus) -> String {
        let margin = status.margin
        if status.closed {
            // "3&2" when decided with holes to spare; closing on the final
            // hole reads "2 UP".
            return status.remaining > 0 ? "\(margin)&\(status.remaining)" : "\(margin) UP"
        }
        if status.holesDecided == 0 {
            return "Not started"
        }
        let thru = status.thruHole.map { " thru \($0)" } ?? ""
        if status.remaining == 0 {
            return margin == 0 ? "Halved" : "Final: \(margin) UP"
        }
        if status.dormieSide != nil {
            return "Dormie \(margin)"
        }
        return margin == 0 ? "AS\(thru)" : "\(margin) UP\(thru)"
    }
}

File: web/src/scoring/matchEngine.ts
import { strokeTable, strokesReceived } from './handicapping';
import type { MatchComputation, MatchHoleResult, MatchStatus } from './evaluation';
import { matchMargin } from './evaluation';
import {
  activePlayers,
  gross,
  participantsOf,
  type RoundSnapshot,
} from './snapshot';
import { holeAt, type HandicapAllowance, type HandicapMode, type MatchSide, type PlayerID } from './types';

/**
 * Pure hole-by-hole match evaluator.
 *
 * Rules encoded here:
 * - **Best ball**: a side's score on a hole is the lowest net score among its
 *   active members. All active members of both sides must have posted before
 *   the hole counts (a partially-entered best-ball hole could otherwise flip
 *   after being "decided").
 * - **Pending holes stay live**: a skipped hole (no scores yet) counts toward
 *   `remaining`, because it can still be filled in. Mathematical closure
 *   (`|up| > remaining`) is therefore safe even with gaps.
 * - **Withdrawals concede**: once every member of a side has withdrawn, each
 *   later hole is conceded to the other side (if it still has anyone standing).
 *   If both sides are gone, remaining holes are halved.
 * - **Auto-close**: once `|up| > remaining` the match is over; later holes are
 *   ignored for this match.
 */

/** Who won one hole, with no notion of match closure. */
export interface RawOutcome {
  /** null = halved. */
  winner: MatchSide | null;
  byConcession: boolean;
}

export interface RawOutcomes {
  /** Decided holes only. */
  byHole: Record<number, RawOutcome>;
  /** Holes with missing scores, ascending. */
  pendingHoles: number[];
}

/**
 * Who won each hole. Nassau computes this once per segment and derives the
 * original bet and every press from the same outcome table.
 */
export function rawOutcomes(
  sideA: PlayerID[],
  sideB: PlayerID[],
  holes: number[],
  snapshot: RoundSnapshot,
  mode: HandicapMode,
  allowance: HandicapAllowance,
): RawOutcomes {
  const participants = participantsOf(snapshot, [...sideA, ...sideB]);
  const courseHoles = snapshot.holeNumbers
    .map((n) => holeAt(snapshot.course, n))
    .filter((h): h is NonNullable<typeof h> => h !== undefined);
  // Strokes fall on the holes where the full round's stroke index puts them —
  // a front-nine segment does not re-spread the handicap.
  const strokes = strokeTable(participants, courseHoles, mode, allowance);

  const net = (player: PlayerID, hole: number): number | undefined => {
    const raw = gross(snapshot, player, hole);
    if (raw === undefined) return undefined;
    return raw - strokesReceived(strokes, player, hole);
  };

  const byHole: Record<number, RawOutcome> = {};
  const pending: number[] = [];

  for (const hole of [...holes].sort((a, b) => a - b)) {
    const activeA = activePlayers(snapshot, sideA, hole);
    const activeB = activePlayers(snapshot, sideB, hole);

    if (activeA.length === 0 && activeB.length === 0) {
      byHole[hole] = { winner: null, byConcession: true };
    } else if (activeA.length === 0) {
      byHole[hole] = { winner: 'b', byConcession: true };
    } else if (activeB.length === 0) {
      byHole[hole] = { winner: 'a', byConcession: true };
    } else {
      const netsA = activeA.map((p) => net(p, hole));
      const netsB = activeB.map((p) => net(p, hole));
      if (netsA.some((n) => n === undefined) || netsB.some((n) => n === undefined)) {
        pending.push(hole);
      } else {
        const bestA = Math.min(...(netsA as number[]));
        const bestB = Math.min(...(netsB as number[]));
        const winner: MatchSide | null = bestA < bestB ? 'a' : bestB < bestA ? 'b' : null;
        byHole[hole] = { winner, byConcession: false };
      }
    }
  }

  return { byHole, pendingHoles: pending };
}

/**
 * Applies cumulative match logic (closure, dormie, display) to a hole range
 * using a precomputed outcome table.
 */
export function matchStatus(holes: number[], outcomes: RawOutcomes): MatchComputation {
  const results: MatchHoleResult[] = [];
  const pending: number[] = [];
  let upA = 0;
  let closed = false;

  const ordered = [...holes].sort((a, b) => a - b);
  for (const hole of ordered) {
    if (closed) break;
    const outcome = outcomes.byHole[hole];
    if (outcome === undefined) {
      pending.push(hole);
      continue;
    }
    if (outcome.winner === 'a') upA += 1;
    else if (outcome.winner === 'b') upA -= 1;
    results.push({
      hole,
      winner: outcome.winner,
      upAAfter: upA,
      byConcession: outcome.byConcession,
    });

    // Undecided = total - decided. Pending holes are still undecided (they may
    // be filled in later), so they count toward what the trailing side could
    // still win.
    const undecidedNow = ordered.length - results.length;
    if (Math.abs(upA) > undecidedNow) {
      closed = true;
    }
  }

  const undecided = ordered.length - results.length;
  const decidedHoles = results.map((r) => r.hole);

  const status: MatchStatus = {
    holesDecided: results.length,
    upA,
    remaining: undecided,
    closed,
    winner: null,
    dormieSide: null,
    thruHole: decidedHoles.length > 0 ? Math.max(...decidedHoles) : null,
    display: '—',
  };

  if (closed) {
    status.winner = upA > 0 ? 'a' : 'b';
  } else if (undecided === 0) {
    status.winner = upA > 0 ? 'a' : upA < 0 ? 'b' : null;
  } else if (Math.abs(upA) === undecided && upA !== 0) {
    status.dormieSide = upA > 0 ? 'a' : 'b';
  }

  status.display = matchDisplay(status);

  return { status, holeResults: results, pendingHoles: pending };
}

/** One-shot convenience: raw outcomes + status over the same range. */
export function computeMatch(
  sideA: PlayerID[],
  sideB: PlayerID[],
  holes: number[],
  snapshot: RoundSnapshot,
  mode: HandicapMode,
  allowance: HandicapAllowance,
): MatchComputation {
  const outcomes = rawOutcomes(sideA, sideB, holes, snapshot, mode, allowance);
  return matchStatus(holes, outcomes);
}

/** Conventional match-play status strings. */
export function matchDisplay(status: MatchStatus): string {
  const margin = matchMargin(status);
  if (status.closed) {
    // "3&2" when decided with holes to spare; closing on the final hole reads
    // "2 UP".
    return status.remaining > 0 ? `${margin}&${status.remaining}` : `${margin} UP`;
  }
  if (status.holesDecided === 0) {
    return 'Not started';
  }
  const thru = status.thruHole !== null ? ` thru ${status.thruHole}` : '';
  if (status.remaining === 0) {
    return margin === 0 ? 'Halved' : `Final: ${margin} UP`;
  }
  if (status.dormieSide !== null) {
    return `Dormie ${margin}`;
  }
  return margin === 0 ? `AS${thru}` : `${margin} UP${thru}`;
}
